{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Llama-3-7B Basic RAG Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "dummy_synth = pd.read_csv('./ollama-3-dummy-responses/ragas_synth.csv')\n",
    "dummy_curated = pd.read_csv('./ollama-3-dummy-responses/ragas_curated.csv')\n",
    "test_synth = pd.read_csv('./ollama-3-test-responses/ragas_synth.csv')\n",
    "test_curated = pd.read_csv('./ollama-3-test-responses/ragas_curated.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>faithfulness</th>\n",
       "      <th>answer_relevancy</th>\n",
       "      <th>answer_correctness</th>\n",
       "      <th>answer_similarity</th>\n",
       "      <th>context_relevancy</th>\n",
       "      <th>context_precision</th>\n",
       "      <th>context_recall</th>\n",
       "      <th>context_entity_recall</th>\n",
       "      <th>context_utilization</th>\n",
       "      <th>harmfulness</th>\n",
       "      <th>correctness</th>\n",
       "      <th>coherence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>14.000000</td>\n",
       "      <td>0.336815</td>\n",
       "      <td>0.497059</td>\n",
       "      <td>0.458162</td>\n",
       "      <td>0.436048</td>\n",
       "      <td>0.219436</td>\n",
       "      <td>0.620690</td>\n",
       "      <td>0.679269</td>\n",
       "      <td>0.031404</td>\n",
       "      <td>0.517241</td>\n",
       "      <td>0.827586</td>\n",
       "      <td>0.896552</td>\n",
       "      <td>0.896552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>8.514693</td>\n",
       "      <td>0.370425</td>\n",
       "      <td>0.233881</td>\n",
       "      <td>0.245461</td>\n",
       "      <td>0.405356</td>\n",
       "      <td>0.150789</td>\n",
       "      <td>0.493804</td>\n",
       "      <td>0.420627</td>\n",
       "      <td>0.099614</td>\n",
       "      <td>0.508548</td>\n",
       "      <td>0.384426</td>\n",
       "      <td>0.309934</td>\n",
       "      <td>0.309934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.024753</td>\n",
       "      <td>-0.099346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>7.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.434009</td>\n",
       "      <td>0.303209</td>\n",
       "      <td>-0.033331</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>14.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.496006</td>\n",
       "      <td>0.480281</td>\n",
       "      <td>0.659498</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>21.000000</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.535104</td>\n",
       "      <td>0.596574</td>\n",
       "      <td>0.769335</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>28.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  faithfulness  answer_relevancy  answer_correctness  \\\n",
       "count   29.000000     29.000000         29.000000           25.000000   \n",
       "mean    14.000000      0.336815          0.497059            0.458162   \n",
       "std      8.514693      0.370425          0.233881            0.245461   \n",
       "min      0.000000      0.000000          0.000000           -0.024753   \n",
       "25%      7.000000      0.000000          0.434009            0.303209   \n",
       "50%     14.000000      0.250000          0.496006            0.480281   \n",
       "75%     21.000000      0.600000          0.535104            0.596574   \n",
       "max     28.000000      1.000000          1.000000            1.000000   \n",
       "\n",
       "       answer_similarity  context_relevancy  context_precision  \\\n",
       "count          29.000000          29.000000          29.000000   \n",
       "mean            0.436048           0.219436           0.620690   \n",
       "std             0.405356           0.150789           0.493804   \n",
       "min            -0.099346           0.000000           0.000000   \n",
       "25%            -0.033331           0.090909           0.000000   \n",
       "50%             0.659498           0.272727           1.000000   \n",
       "75%             0.769335           0.272727           1.000000   \n",
       "max             1.000000           0.636364           1.000000   \n",
       "\n",
       "       context_recall  context_entity_recall  context_utilization  \\\n",
       "count       29.000000              29.000000            29.000000   \n",
       "mean         0.679269               0.031404             0.517241   \n",
       "std          0.420627               0.099614             0.508548   \n",
       "min          0.000000               0.000000             0.000000   \n",
       "25%          0.500000               0.000000             0.000000   \n",
       "50%          1.000000               0.000000             1.000000   \n",
       "75%          1.000000               0.000000             1.000000   \n",
       "max          1.000000               0.500000             1.000000   \n",
       "\n",
       "       harmfulness  correctness  coherence  \n",
       "count    29.000000    29.000000  29.000000  \n",
       "mean      0.827586     0.896552   0.896552  \n",
       "std       0.384426     0.309934   0.309934  \n",
       "min       0.000000     0.000000   0.000000  \n",
       "25%       1.000000     1.000000   1.000000  \n",
       "50%       1.000000     1.000000   1.000000  \n",
       "75%       1.000000     1.000000   1.000000  \n",
       "max       1.000000     1.000000   1.000000  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dummy_synth.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>faithfulness</th>\n",
       "      <th>answer_relevancy</th>\n",
       "      <th>answer_correctness</th>\n",
       "      <th>answer_similarity</th>\n",
       "      <th>context_relevancy</th>\n",
       "      <th>context_precision</th>\n",
       "      <th>context_recall</th>\n",
       "      <th>context_entity_recall</th>\n",
       "      <th>context_utilization</th>\n",
       "      <th>harmfulness</th>\n",
       "      <th>correctness</th>\n",
       "      <th>coherence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>29.000000</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>23.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>14.000000</td>\n",
       "      <td>0.303968</td>\n",
       "      <td>0.443376</td>\n",
       "      <td>0.473278</td>\n",
       "      <td>0.402347</td>\n",
       "      <td>0.031532</td>\n",
       "      <td>0.455412</td>\n",
       "      <td>0.632857</td>\n",
       "      <td>0.014368</td>\n",
       "      <td>0.451964</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.782609</td>\n",
       "      <td>0.739130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>8.514693</td>\n",
       "      <td>0.311692</td>\n",
       "      <td>0.234651</td>\n",
       "      <td>0.256559</td>\n",
       "      <td>0.395234</td>\n",
       "      <td>0.024600</td>\n",
       "      <td>0.390691</td>\n",
       "      <td>0.461093</td>\n",
       "      <td>0.054867</td>\n",
       "      <td>0.367300</td>\n",
       "      <td>0.428932</td>\n",
       "      <td>0.421741</td>\n",
       "      <td>0.448978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.025206</td>\n",
       "      <td>-0.123094</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>7.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.408075</td>\n",
       "      <td>0.354854</td>\n",
       "      <td>-0.050175</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>14.000000</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.498855</td>\n",
       "      <td>0.432561</td>\n",
       "      <td>0.502015</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>21.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.588848</td>\n",
       "      <td>0.592003</td>\n",
       "      <td>0.749998</td>\n",
       "      <td>0.050847</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>28.000000</td>\n",
       "      <td>0.850000</td>\n",
       "      <td>0.869517</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.094828</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  faithfulness  answer_relevancy  answer_correctness  \\\n",
       "count   29.000000     21.000000         23.000000           28.000000   \n",
       "mean    14.000000      0.303968          0.443376            0.473278   \n",
       "std      8.514693      0.311692          0.234651            0.256559   \n",
       "min      0.000000      0.000000          0.000000           -0.025206   \n",
       "25%      7.000000      0.000000          0.408075            0.354854   \n",
       "50%     14.000000      0.266667          0.498855            0.432561   \n",
       "75%     21.000000      0.500000          0.588848            0.592003   \n",
       "max     28.000000      0.850000          0.869517            1.000000   \n",
       "\n",
       "       answer_similarity  context_relevancy  context_precision  \\\n",
       "count          29.000000          29.000000          29.000000   \n",
       "mean            0.402347           0.031532           0.455412   \n",
       "std             0.395234           0.024600           0.390691   \n",
       "min            -0.123094           0.000000           0.000000   \n",
       "25%            -0.050175           0.014706           0.000000   \n",
       "50%             0.502015           0.017857           0.500000   \n",
       "75%             0.749998           0.050847           0.833333   \n",
       "max             1.000000           0.094828           1.000000   \n",
       "\n",
       "       context_recall  context_entity_recall  context_utilization  \\\n",
       "count       20.000000              29.000000            29.000000   \n",
       "mean         0.632857               0.014368             0.451964   \n",
       "std          0.461093               0.054867             0.367300   \n",
       "min          0.000000               0.000000             0.000000   \n",
       "25%          0.037500               0.000000             0.000000   \n",
       "50%          1.000000               0.000000             0.583333   \n",
       "75%          1.000000               0.000000             0.700000   \n",
       "max          1.000000               0.250000             1.000000   \n",
       "\n",
       "       harmfulness  correctness  coherence  \n",
       "count    22.000000    23.000000  23.000000  \n",
       "mean      0.772727     0.782609   0.739130  \n",
       "std       0.428932     0.421741   0.448978  \n",
       "min       0.000000     0.000000   0.000000  \n",
       "25%       1.000000     1.000000   0.500000  \n",
       "50%       1.000000     1.000000   1.000000  \n",
       "75%       1.000000     1.000000   1.000000  \n",
       "max       1.000000     1.000000   1.000000  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_synth.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>faithfulness</th>\n",
       "      <th>answer_relevancy</th>\n",
       "      <th>answer_correctness</th>\n",
       "      <th>answer_similarity</th>\n",
       "      <th>context_relevancy</th>\n",
       "      <th>context_precision</th>\n",
       "      <th>context_recall</th>\n",
       "      <th>context_entity_recall</th>\n",
       "      <th>context_utilization</th>\n",
       "      <th>harmfulness</th>\n",
       "      <th>correctness</th>\n",
       "      <th>coherence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>60.000000</td>\n",
       "      <td>59.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>59.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>29.500000</td>\n",
       "      <td>0.335876</td>\n",
       "      <td>0.591683</td>\n",
       "      <td>0.415959</td>\n",
       "      <td>0.729319</td>\n",
       "      <td>0.262121</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.400217</td>\n",
       "      <td>0.031498</td>\n",
       "      <td>0.866667</td>\n",
       "      <td>0.915254</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.966667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>17.464249</td>\n",
       "      <td>0.334936</td>\n",
       "      <td>0.068120</td>\n",
       "      <td>0.110587</td>\n",
       "      <td>0.103881</td>\n",
       "      <td>0.107293</td>\n",
       "      <td>0.278718</td>\n",
       "      <td>0.303373</td>\n",
       "      <td>0.067405</td>\n",
       "      <td>0.342803</td>\n",
       "      <td>0.280894</td>\n",
       "      <td>0.251549</td>\n",
       "      <td>0.181020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.419199</td>\n",
       "      <td>0.230787</td>\n",
       "      <td>0.424720</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>14.750000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.543871</td>\n",
       "      <td>0.335309</td>\n",
       "      <td>0.668827</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.193478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>29.500000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.586896</td>\n",
       "      <td>0.394719</td>\n",
       "      <td>0.752439</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>44.250000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.640197</td>\n",
       "      <td>0.462946</td>\n",
       "      <td>0.798770</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.011905</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>59.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.745200</td>\n",
       "      <td>0.953894</td>\n",
       "      <td>0.888476</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  faithfulness  answer_relevancy  answer_correctness  \\\n",
       "count   60.000000     59.000000         60.000000           55.000000   \n",
       "mean    29.500000      0.335876          0.591683            0.415959   \n",
       "std     17.464249      0.334936          0.068120            0.110587   \n",
       "min      0.000000      0.000000          0.419199            0.230787   \n",
       "25%     14.750000      0.000000          0.543871            0.335309   \n",
       "50%     29.500000      0.333333          0.586896            0.394719   \n",
       "75%     44.250000      0.500000          0.640197            0.462946   \n",
       "max     59.000000      1.000000          0.745200            0.953894   \n",
       "\n",
       "       answer_similarity  context_relevancy  context_precision  \\\n",
       "count          60.000000          60.000000          60.000000   \n",
       "mean            0.729319           0.262121           0.916667   \n",
       "std             0.103881           0.107293           0.278718   \n",
       "min             0.424720           0.000000           0.000000   \n",
       "25%             0.668827           0.181818           1.000000   \n",
       "50%             0.752439           0.272727           1.000000   \n",
       "75%             0.798770           0.272727           1.000000   \n",
       "max             0.888476           0.545455           1.000000   \n",
       "\n",
       "       context_recall  context_entity_recall  context_utilization  \\\n",
       "count       60.000000              60.000000            60.000000   \n",
       "mean         0.400217               0.031498             0.866667   \n",
       "std          0.303373               0.067405             0.342803   \n",
       "min          0.000000               0.000000             0.000000   \n",
       "25%          0.193478               0.000000             1.000000   \n",
       "50%          0.366667               0.000000             1.000000   \n",
       "75%          0.600000               0.011905             1.000000   \n",
       "max          1.000000               0.375000             1.000000   \n",
       "\n",
       "       harmfulness  correctness  coherence  \n",
       "count    59.000000    60.000000  60.000000  \n",
       "mean      0.915254     0.933333   0.966667  \n",
       "std       0.280894     0.251549   0.181020  \n",
       "min       0.000000     0.000000   0.000000  \n",
       "25%       1.000000     1.000000   1.000000  \n",
       "50%       1.000000     1.000000   1.000000  \n",
       "75%       1.000000     1.000000   1.000000  \n",
       "max       1.000000     1.000000   1.000000  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dummy_curated.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>faithfulness</th>\n",
       "      <th>answer_relevancy</th>\n",
       "      <th>answer_correctness</th>\n",
       "      <th>answer_similarity</th>\n",
       "      <th>context_relevancy</th>\n",
       "      <th>context_precision</th>\n",
       "      <th>context_recall</th>\n",
       "      <th>context_entity_recall</th>\n",
       "      <th>context_utilization</th>\n",
       "      <th>harmfulness</th>\n",
       "      <th>correctness</th>\n",
       "      <th>coherence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>60.000000</td>\n",
       "      <td>53.000000</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>55.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>29.500000</td>\n",
       "      <td>0.245650</td>\n",
       "      <td>0.544207</td>\n",
       "      <td>0.426810</td>\n",
       "      <td>0.701981</td>\n",
       "      <td>0.015713</td>\n",
       "      <td>0.577940</td>\n",
       "      <td>0.307460</td>\n",
       "      <td>0.010198</td>\n",
       "      <td>0.667569</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.981818</td>\n",
       "      <td>0.981818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>17.464249</td>\n",
       "      <td>0.333516</td>\n",
       "      <td>0.167732</td>\n",
       "      <td>0.134944</td>\n",
       "      <td>0.105308</td>\n",
       "      <td>0.012480</td>\n",
       "      <td>0.365274</td>\n",
       "      <td>0.262297</td>\n",
       "      <td>0.043714</td>\n",
       "      <td>0.339127</td>\n",
       "      <td>0.290129</td>\n",
       "      <td>0.134840</td>\n",
       "      <td>0.134840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153669</td>\n",
       "      <td>0.426075</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>14.750000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.522566</td>\n",
       "      <td>0.319198</td>\n",
       "      <td>0.637797</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>0.237500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.404167</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>29.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.577545</td>\n",
       "      <td>0.403790</td>\n",
       "      <td>0.706731</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.309524</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.779861</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>44.250000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.626235</td>\n",
       "      <td>0.509780</td>\n",
       "      <td>0.790779</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.925000</td>\n",
       "      <td>0.553571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>59.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.766871</td>\n",
       "      <td>0.803176</td>\n",
       "      <td>0.867822</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  faithfulness  answer_relevancy  answer_correctness  \\\n",
       "count   60.000000     53.000000         55.000000           55.000000   \n",
       "mean    29.500000      0.245650          0.544207            0.426810   \n",
       "std     17.464249      0.333516          0.167732            0.134944   \n",
       "min      0.000000      0.000000          0.000000            0.153669   \n",
       "25%     14.750000      0.000000          0.522566            0.319198   \n",
       "50%     29.500000      0.000000          0.577545            0.403790   \n",
       "75%     44.250000      0.500000          0.626235            0.509780   \n",
       "max     59.000000      1.000000          0.766871            0.803176   \n",
       "\n",
       "       answer_similarity  context_relevancy  context_precision  \\\n",
       "count          60.000000          60.000000          60.000000   \n",
       "mean            0.701981           0.015713           0.577940   \n",
       "std             0.105308           0.012480           0.365274   \n",
       "min             0.426075           0.000000           0.000000   \n",
       "25%             0.637797           0.012658           0.237500   \n",
       "50%             0.706731           0.015625           0.700000   \n",
       "75%             0.790779           0.015625           0.925000   \n",
       "max             0.867822           0.062500           1.000000   \n",
       "\n",
       "       context_recall  context_entity_recall  context_utilization  \\\n",
       "count       30.000000              60.000000            60.000000   \n",
       "mean         0.307460               0.010198             0.667569   \n",
       "std          0.262297               0.043714             0.339127   \n",
       "min          0.000000               0.000000             0.000000   \n",
       "25%          0.000000               0.000000             0.404167   \n",
       "50%          0.309524               0.000000             0.779861   \n",
       "75%          0.553571               0.000000             0.950000   \n",
       "max          0.800000               0.285714             1.000000   \n",
       "\n",
       "       harmfulness  correctness  coherence  \n",
       "count    55.000000    55.000000  55.000000  \n",
       "mean      0.909091     0.981818   0.981818  \n",
       "std       0.290129     0.134840   0.134840  \n",
       "min       0.000000     0.000000   0.000000  \n",
       "25%       1.000000     1.000000   1.000000  \n",
       "50%       1.000000     1.000000   1.000000  \n",
       "75%       1.000000     1.000000   1.000000  \n",
       "max       1.000000     1.000000   1.000000  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_curated.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['Glycans favoring cell motility (e.g., polysialic acid) will be favored for development but become detrimental when accidentally exploited by malignant cells. Glycans on reproductive tract secretions favoring survival of sperm (e.g., Glycodelin S) might be counter-selected in females who benefit from a different glycan form (Glycodelin A), which challenges male gametes as part of female quality control. Unique glycans evolved as reliable SAMPs can become a liability if exploited by pathogens through molecular mimicry. Given the rapid evolution of extrinsic pathogens and their frequent use of glycans as targets for host recognition, it seems likely that a significant portion of the overall diversity in vertebrate cell-surface glycan structure reflects such pathogen-mediated selection processes. Meanwhile, even one critical intrinsic role of a glycan could disallow its elimination as a mechanism to evade pathogens. Thus, glycan expression patterns may represent trade-offs between evading pathogens (or accommodating symbionts) and preserving intrinsic functions. More gene disruption studies in intact animals could help differentiate intrinsic and extrinsic glycan functions. More systematic comparative glycobiology could also contribute, making predictions about intrinsic glycan function—that is, the consistent (conserved) expression of the same structure in the same cell type across several taxa would imply a critical intrinsic role. Such work might also help define the rate of glycan diversification during evolution, better define the relative roles of the intrinsic and extrinsic selective forces, and eventually lead to a better understanding of the functional significance of glycan diversification during evolution. The possibility that pathogen-driven glycan diversification might even favor the process of sympatric speciation (via reproductive isolation) also needs to be further explored. The authors appreciate helpful comments and suggestions from Cristina De Castro and Christopher Mark West.'\\n 'Adhesin–glycan binding triggers invasion of the merozoites into red blood cells, where they develop into mature schizonts that rupture and release newly formed merozoites into the bloodstream. Many commonly used clinical antimalarial medications, such as chloroquine, target the parasite during this erythrocytic asexual reproduction stage. Various glycan-binding proteins mediate adhesion of microorganisms to host cells or tissues. These interactions are often a prerequisite for infection or symbiosis, with binding-deficient mutants being unable to initiate or maintain these relationships. Interestingly, glycans recognized by microbial surface lectins have been shown to block the adhesion of bacteria to animal cells in vitro and in vivo, and thus may protect animals against infection by such microorganisms. For example, coadministration of methyl α-mannoside with type-1-fimbriated E. coli into the bladder of mice can reduce microbial burden in a mouse model of urinary tract infection, whereas methyl α-glucoside, which does not bind to FimH, has no effect. Furthermore, lacto-N-neotetraose (LNnT) and its α2-3- and α2-6-sialylated derivatives block the adherence of Streptococcus pneumoniae to respiratory epithelial cells in vitro. In addition, these glycans prevent colonization of the nasopharynx and attenuate the course of pneumonia in rodent models of pneumococcal infection. Naturally occurring host glycans can also serve protective roles. For example, secreted mucins and human milk oligosaccharides (HMOs) act as natural receptor decoys, protecting the host from pathogens and shaping the microbiota composition. Exogenous heparin and structurally related polysaccharides are known to inhibit viral replication, suggesting a potential approach for the development of polysaccharide-based antiviral pharmaceutical agents. For example, heparin octasaccharide decoy liposomes were recently shown to inhibit the replication of numerous viruses including HSV and respiratory syncytial virus. As more crystal structures are elucidated, the ability to design small molecule inhibitors that fit into the carbohydrate-recognition domains of adhesins should improve. Already, the structures of influenza hemagglutinin and sialidase have suggested numerous ways to modify sialic acid to fit better into the active sites. Some of these compounds are presently in clinical use to limit the spread of influenza. The authors acknowledge contributions to the previous version of this chapter from Jeffrey D. Esko, Victor Nizet, and the late Nathan Sharon and helpful comments and suggestions from Frederique Lisacek and Ajit Varki.'\\n 'Information about these is available in Further Reading. With the development of direct electron detectors, cryo-electron microscopy (cryo-EM) has become one of the most powerful techniques to obtain high-resolution structural information on biological macromolecules. With reported resolutions exceeding 2 Å, many molecular details are revealed under native-like conditions. By eliminating the need for growing well-ordered three-dimensional (3D) crystals and operating at fairly low sample concentrations, cryo-EM can provide atomic level insights into many biological samples, from soluble and membrane-integrated protein complexes to filamentous polymers and entire viruses. Cryo-EM can be roughly divided into two main directions: one working with single (usually purified) particles, single-particle analysis (SPA), and the other using a tomography approach to analyze species in larger assemblies, such as in vitro–assembled scaffolds or even native cells and tissues. For SPA, samples are generally analyzed in a thin layer of vitreous ice containing the particles in random orientations. Data is collected in the form of movies, which allows correction for beam-induced drift, followed by estimation of the contrast transfer function for each micrograph; both are necessary to obtain near-atomic resolutions. The individual particles are then computationally extracted from the micrographs, sorted, and ultimately aligned in three dimensions to reconstruct the molecular structure. For cryo-electron tomography of, for example, a vitrified cell or virus particle, a tilt series of images is acquired to obtain different “specimen views” necessary for 3D reconstruction. This technique continues to face technical challenges, in part because of limitations on the tilt angles that can be achieved. However, it is a powerful tool to image, for example, the glycocalyx of various tissues, plant and fungal cells walls, or microbial cell envelopes and capsules. A major advantage of cryo-EM is that sample heterogeneity and/or conformational flexibility does not preclude analyses. To obtain well-diffracting crystals of a glycosylated protein, for example, the conformationally heterogenous glycans are often removed enzymatically to facilitate crystallization. For cryo-EM, these pretreatments are generally unnecessary, thereby providing molecular details of proteins in the context of posttranslational modifications. Analyses of fully glycosylated viral envelope proteins are fascinating examples that document the potential of cryo-EM for studying protein–carbohydrate interactions (Figure 30.5). Other examples include polysaccharide-synthesizing enzymes bound to their polymeric products as well as integral membrane transporters associated with lipopolysaccharide substrates. Cryo-electron microscopy structure of the native fully glycosylated HIV-1 envelope trimer. Protein subunits are shown as gray surfaces and glycans are shown as ball and sticks (PDB entry 5FUU). Carbohydrates are colored green and pale green (β- (more...) We can thus look forward to correlating unprecedented structural insights on protein glycosylation and complex carbohydrate interactions with biochemical, functional, spectroscopic, and in silico approaches in the coming years. Experimental structural information obtained by crystallographic, NMR, and cryo-EM methods have clearly been of value in building an understanding of the molecular interactions that lead to glycan recognition by proteins. However, systems in which interactions are of interest far outnumber the cases in which these methods can be applied. Most crystal structures contain either small ligands or yield useful electron densities for only parts of larger ligands. The same is true to a certain extent for cryo-EM structures. NMR methods, although giving detailed information on bound ligand geometries, frequently give only qualitative information on parts of ligands or protein that are in intimate contact with each other. All three methods require substantial effort, particularly in preparing samples for investigation. A particular problem for glycans of interest is that they are often complex molecules that are difficult to prepare in highly pure forms or in the quantities needed for experimental investigation. There are also functionally important dynamic processes (e.g., enzymatic conversions of glycan substrates to products and transport of glycans) that are not well-represented by static, thermodynamically stable structures. Computational methods can extend analyses into these less accessible regions of structural investigation. Computational contributions to our understanding of glycan properties have a long history, beginning with a very fundamental understanding of factors influencing anomeric configuration and glycosidic torsion angles. These glycan-specific factors, such as the anomeric effect and the exo-anomeric effect, are described more thoroughly in Chapters 2, 3, and 50.'\\n 'The gel is then stained with Coomassie Brilliant Blue and an immunoelectrophoretogram is obtained. Glycoproteins not interacting with the GBP or antibody have faster mobility than the complex. The amount of glycoprotein or ligand is determined by the area under the curves obtained in the second dimensional analysis. This method is useful for studying glycoforms of proteins and has been particularly valuable in analyzing glycoforms of α1-acid glycoprotein (an acute-phase glycoprotein) in serum and changes in its α1-3 fucosylation. An indirect approach to studying protein–glycan interactions is to express a copy DNA (cDNA) encoding a glycosyltransferase in an animal or bacterial cell (Chapter 56). The adhesion of the modified cell (either transiently or stably transfected) to a GBP or antibody is then measured and taken to reflect the binding of the GBP or antibody to the new glycans (neoglycans) on the cell surface. Conversely, a cDNA encoding a GBP is expressed in cells and their ability to bind glycan ligands is tested. The expression of selectins, Siglecs, and other GBPs on the cell surface of transfected cells has been helpful in evaluating the roles of GBPs in cell adhesion under physiological flow conditions.'\\n 'For example, the name of the closed ring Neu5Ac is 5-acetamido-3,5-dideoxy-D-glycero-D-galacto-non-2-ulopyranosonic acid. Beyond the currently known core units (Figure 15.1), additional substitutions are designated by letter codes (Ac, acetyl; Gc, glycolyl; Me, methyl; Lt, lactyl; and S, sulfate), and these are listed along with numbers indicating their location relative to the carbons (e.g., 9-O-acetyl-8-O-methyl-N-acetylneuraminic acid is Neu5,9Ac28Me). Among NulOs as a whole, there are now at least 138 known structural variants. If one is uncertain of the exact type of molecule present, the generic abbreviations Sia or NulO should be used. Partial structural information can also be incorporated. For example, a Sia of otherwise unknown type with an O-acetyl substitution at the C-9 position could be written as Sia9Ac. With regard to the bacterial NulOs, the names Leg, Pse, 4eLeg, 8eLeg, and Aci are often used synonymously with the 5,7-di-N-acetylated forms of these molecules (Figure 15.1). To be consistent with Sia nomenclature (as with Neu) and to avoid confusion, it is recommended that the substituted forms be specified whenever possible (e.g., Leg5,7Ac2). Free amino forms of the prokaryotic NulOs have not to our knowledge been reported in nature. Instead, many different types of substitutions of the C-5 and C-7 positions have been reported for the exclusively prokaryotic NulOs, including acetyl (Ac), formyl (Fo), (R)-hydroxybutyryl (3RHb), (S)-3-hydroxybutyryl (3SHb), 4-hydroxybutyryl (4Hb), 3,4-dihydroxybutyryl (3,4Hb), acetimidoyl (Am), N-methyl-acetimidoyl (AmMe), methyl (Me), D-alanyl (Ala), N-acetyl-D-alanyl (AlaNAc), N-methyl-5-glutamyl (GluNMe), L-glyceryl (Gr), and/or 2,3-di-O-methyl-glyceryl (Me2Gr) groups. Additional variety derives from acetyl groups at O4 and O8, as well as acetyl, N-acetyl-glutaminyl (GlnNAc), and glycyl (Gly) substitutions at O8. The authors acknowledge contributions to previous versions of this chapter by the late Roland Schauer and appreciate helpful comments and suggestions from Hamed Jafar-Nejad.']\""
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_curated['contexts'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_context_list(sample):\n",
    "    \"\"\"converts the contexts column into a list of strings\"\"\"\n",
    "    sample = sample.replace('[', '')\n",
    "    sample = sample.replace(']', '')\n",
    "    sample = sample.split('\\n')\n",
    "    for i, t in enumerate(sample):\n",
    "        sample[i] = t[1:-1]\n",
    "        if sample[i][0] == \"'\":\n",
    "            sample[i] = sample[i][1:]\n",
    "\n",
    "    return sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "from nltk import word_tokenize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2095"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample = test_curated['contexts'][0]\n",
    "sample = extract_context_list(sample)\n",
    "lengths = [len(word_tokenize(i)) for i in sample]\n",
    "sum(lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "392.0833333333333\n"
     ]
    }
   ],
   "source": [
    "context_sizes = []\n",
    "for retrieved in test_curated['contexts']:\n",
    "    extracted = extract_context_list(retrieved)\n",
    "    for text in extracted:\n",
    "        size = len(word_tokenize(text))\n",
    "        context_sizes.append(size)\n",
    "\n",
    "mean_context_size = sum(context_sizes) / len(context_sizes)\n",
    "print(mean_context_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "778"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(context_sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "114"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min(context_sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(word_tokenize(test_synth['question'][13]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "glybot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
